runtime/mux.go

```diff
@@ -36,12 +39,37 @@ func WithForwardResponseOption(forwardResponseOption func(context.Context, http.
 	}
 }
 
+// HeaderMatcherFunc checks whether a header key should be forwarded to/from gRPC context.
+type HeaderMatcherFunc func(string) (string, bool)
+
+// WithHeaderMatcher returns a ServeMuxOption representing the headerMatcher.
+//
+// This matcher will be called with each header in http.Request. If matcher returns true, that header will be
+// passed to gRPC context. To transform the header before passing to gRPC context, matcher should return modified header.
+func WithHeaderMatcher(fn HeaderMatcherFunc) ServeMuxOption {
+	return func(mux *ServeMux) {
+		mux.headerMatchers = append(mux.headerMatchers, fn)
+	}
+}
+
+// WithMetadata returns a ServeMuxOption for passing metadata to a gRPC context.
+//
+// This can be used by services that need to read from http.Request and modify gRPC context. A common use case
+// is reading token from cookie and adding it in gRPC context.
+func WithMetadata(annotator func(context.Context, *http.Request) metadata.MD) ServeMuxOption {
+	return func(serveMux *ServeMux) {
+		serveMux.metadataAnnotators = append(serveMux.metadataAnnotators, annotator)
+	}
+}
+
 // NewServeMux returns a new ServeMux whose internal mapping is empty.
 func NewServeMux(opts ...ServeMuxOption) *ServeMux {
 	serveMux := &ServeMux{
 		handlers:               make(map[string][]handler),
 		forwardResponseOptions: make([]func(context.Context, http.ResponseWriter, proto.Message) error, 0),
 		marshalers:             makeMarshalerMIMERegistry(),
+		headerMatchers:         make([]HeaderMatcherFunc, 0),
```

Comment:
There are 2 separate use-cases:

- headerMatchers: Whether to pass header from http request to gRPC context (with optional transformation). This is either 1->1 or 1->0.

- metadataAnnotator: This is used to add additional metdata to gRPC context. This is 0->1 . This can't be done using header matcher.

Here is code snippet from our prod server:
```
gwMux := gwrt.NewServeMux(
	gwrt.EqualFoldMatcher("Cookie"),
	gwrt.PrefixFoldMatcher("x-csrf-"),
	gwrt.PrefixFoldMatcher("access-control-"),
	gwrt.WithMetadata(func(c context.Context, req *http.Request) metadata.MD {
		return metadata.Pairs("x-forwarded-method", req.Method)
	}))
```